### PR TITLE
index op with int32 support

### PR DESCRIPTION
**Pull request by jianyuh**
Differential Revision: D40089960

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Looks good to me, but just curious why

**Comment by jianyuh** [comment]
> Looks good to me, but just curious why

Thanks for approving! See more context in https://github.com/pytorch/pytorch/pull/86309 . I will wait for @ngimel 's signal to check if this change is good.

File: aten/src/ATen/native/IndexingUtils.h [hunk 1]
```diff
@@ -52,8 +52,8 @@ static C10_UNUSED void checkIndexTensorTypes(IOptTensorListRef indices) {
   for (const auto& tensor : indices) {
     if (tensor.has_value() && tensor->defined()) {
       auto scalarType = tensor->scalar_type();
-      if (scalarType != kLong && scalarType != kByte && scalarType != kBool) {
-          TORCH_CHECK_INDEX(false, "tensors used as indices must be long, byte or bool tensors");
+      if (scalarType != kLong && scalarType != kInt && scalarType != kByte && scalarType != kBool) {
```
Comment (ngimel) [inline aten/src/ATen/native/IndexingUtils.h:55]: can you universally expand the types accepted by this function? It's called from many places, some of which (maybe) don't accept int32 indices.

**Comment by ngimel** [comment]
Wait, but `index_select` already works with int32 today? @qizzzh enabled this some time back.

File: aten/src/ATen/native/IndexingUtils.h (hunk 1 above)
Comment (jianyuh) [reply aten/src/ATen/native/IndexingUtils.h]: https://github.com/pytorch/pytorch/pull/86309 is more reasonable approach. Was trying to check if we can directly remove int constraint but looks like it fails even for CPU index ops.

**Comment by jianyuh** [comment]
> Wait, but `index_select` already works with int32 today? @qizzzh enabled this some time back.

Sorry I meant to check index op.

**Pull request closed without merging**